test/performance/odp_dmafwd.c

```diff
@@ -0,0 +1,973 @@
+			ODPH_ERR("Error creating DMA session\n");
+			return false;
+		}
+
+		odp_dma_pool_param_init(&compl_pool_param);
+		compl_pool_param.num = config->num_pkts;
+		thr->compl_pool = odp_dma_pool_create(PROG_NAME "_dma_compl", &compl_pool_param);
+
+		if (thr->compl_pool == ODP_POOL_INVALID) {
+			ODPH_ERR("Error creating DMA event completion pool\n");
+			return false;
+		}
+
+		thr->copy_pool = config->copy_pool;
+		thr->trs_pool = config->trs_pool;
+		odp_queue_param_init(&queue_param);
+		queue_param.type = ODP_QUEUE_TYPE_SCHED;
+		queue_param.sched.sync = ODP_SCHED_SYNC_PARALLEL;
+		queue_param.sched.prio = odp_schedule_max_prio();
+		thr->compl_q = odp_queue_create(PROG_NAME "_dma_compl", &queue_param);
+
+		if (thr->compl_q == ODP_QUEUE_INVALID) {
+			ODPH_ERR("Error creating DMA completion queue\n");
+			return false;
+		}
+	}
+
+	config->copy_fn = dma_copy_and_send_packets;
+
+	return true;
+}
+
+static odp_bool_t setup_pktios(prog_config_t *config)
+{
+	odp_pool_param_t pool_param;
+	pktio_t *pktio;
+	odp_pktio_param_t pktio_param;
+	odp_pktin_queue_param_t pktin_param;
+	odp_pktout_queue_param_t pktout_param;
+
+	odp_pool_param_init(&pool_param);
+	pool_param.pkt.seg_len = config->pkt_len;
+	pool_param.pkt.len = config->pkt_len;
+	pool_param.pkt.num = config->num_pkts;
+	pool_param.type = ODP_POOL_PACKET;
+	config->pktio_pool = odp_pool_create(PROG_NAME, &pool_param);
+
+	if (config->pktio_pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating packet I/O pool\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_ifs; ++i) {
+		pktio = &config->pktios[i];
+		odp_pktio_param_init(&pktio_param);
+		pktio_param.in_mode = ODP_PKTIN_MODE_SCHED;
+		pktio_param.out_mode = ODP_PKTOUT_MODE_DIRECT;
+		pktio->handle = odp_pktio_open(pktio->name, config->pktio_pool, &pktio_param);
+
+		if (pktio->handle == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Error opening packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktin_queue_param_init(&pktin_param);
+		pktin_param.queue_param.sched.prio = odp_schedule_default_prio();
+
+		if (odp_pktin_queue_config(pktio->handle, &pktin_param) < 0) {
+			ODPH_ERR("Error configuring packet I/O input queues (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktout_queue_param_init(&pktout_param);
+
+		if (odp_pktout_queue_config(pktio->handle, &pktout_param) < 0) {
+			ODPH_ERR("Error configuring packet I/O output queues (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (odp_pktout_queue(pktio->handle, &pktio->out_q, 1) != 1) {
+			ODPH_ERR("Error querying packet I/O output queue (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (odp_pktio_start(pktio->handle) < 0) {
+			ODPH_ERR("Error starting packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void drain(void)
+{
+	odp_event_t ev;
+	odp_event_type_t type;
+	odp_dma_result_t res;
+	odp_buffer_t buf;
+	transfer_t *trs;
+
+	while (true) {
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
```

Comment:
Better to add some timeout here in case there are still running DMA operations?

Reply:
Added a couple of seconds in v2.